Adopt the Extended Heat Index calculation in zone resilience

yujiex opened the pull request:
Pull request overview
---------------------
 - Fixes #10538
 - This PR implements the extended HI. The code is adapted from python code on [this page](https://romps.berkeley.edu/papers/pubs-2020-heatindex.html)

### Introduction

The heat index reflects human perceived hotness considering both temperature and humidity. The current heat index calculation in EnergyPlus is based on Steadman’s method, which gives nonphysical results in extreme weather conditions. The [extended heat index](https://journals.ametsoc.org/view/journals/apme/61/10/JAMC-D-22-0021.1.xml) fixes this issue and outputs a more realistic heat index. It can also be associated with the health outcomes (heat stroke, etc.) of an idealized human, helpful in the quantification of damages from extreme weather.

The following plot shows the absolute and percent difference between the extended HI implemented in this feature branch and the HI in the develop branch

![absolute_diff_newEP_old](https://github.com/NREL/EnergyPlus/assets/13421161/85374c11-e0dd-4a6e-96ce-16249d49d268)

![percent_diff_newEP_old](https://github.com/NREL/EnergyPlus/assets/13421161/cd9d01aa-b297-4d8d-b7bc-152b145f290d)

The difference are mainly in regions with higher temperature and humidity. See the following heat index plot from the [National Weather Service Website](https://www.weather.gov/ama/heatindex)

![image](https://github.com/NREL/EnergyPlus/assets/13421161/1d5a3d5f-c5a1-400c-a79a-d426c75d53a6)

The following is the extended heat index implemented in this feature branch.

![extendedHI_EP](https://github.com/NREL/EnergyPlus/assets/13421161/7fd1fb5a-c9f1-44b2-aab7-d0b2b1f58016)

There are also some slight differences between the extended HI implemented in this feature and the heatindex.py results by Lu and Romps. The maximum difference is around e-5. This difference is originated from the root solver difference between EnergyPlus and heatindex.py

![absolute_diff_newEP_new](https://github.com/NREL/EnergyPlus/assets/13421161/0589a797-bfaa-4d51-b43d-833fe69bf012)

![percent_diff_newEP_new](https://github.com/NREL/EnergyPlus/assets/13421161/91cfdedb-67a6-47f2-b67c-f945df8856d6)

### Regression diffs

The following files have regression diffs. The ESO diffs all happen in the "ZONE ONE:Zone Heat Index [C]" variable. The table diffs are in heat index hours. Note that the heat index hours counts the number of hours with heat index in a certain range. These changes are expected as this feature changes the heat index calculation method.
 
![image](https://github.com/user-attachments/assets/dfe9839f-2cf8-4f59-803d-30982c3395bc)

Using "1ZoneUncontrolled_Win_ASH55_Thermal_Comfort.idf" as an example, the diff looks like the following

![image](https://github.com/NREL/EnergyPlus/assets/13421161/c266db85-a84a-41da-915d-79a4873fae7c)

RH is 100 or close to 100. When temperature is in the range of 90F to 100F, the HI computed using the Steadman and the extended heat index method can differ by 0 to 12C. 

Some other diffs are in resilience report summary tables. These tables are produced when "AllSummary" or "ThermalResilienceSummary" are present in "Output:Table:SummaryReports". For example, this is the diff of test file "_5ZoneAirCooled_LeapYear_annual.idf". The diffs are due to the change in the heat index calculation in this feature branch.
![image](https://github.com/NREL/EnergyPlus/assets/13421161/13e3ca17-151a-459b-a25a-bef9dbd26c2c)

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
Initial commit

Commit:
add license info

Commit:
add extended HI as a new output for testing, will replace old

Commit:
fix error in main function and a typo

Commit:
add unit test, change root solver algorithm type

Commit:
separate out RH=0 test case in find_T, still has issue

Commit:
Merge remote-tracking branch 'origin/develop' into extendedHI

Commit:
use extended HI to replace old HI, no new var created

Commit:
add a test function to compare old and new HI

in TEST_F(EnergyPlusFixture, extendedHI_heatindex_compare), the HI_values used
here as a baseline is output from the heatindex.py from here
https://romps.berkeley.edu/papers/pubs-2020-heatindex.html

Commit:
minor: double to Real64

Commit:
Add documentation

Commit:
use constexpr for constants

Commit:
clang-format

Commit:
Fix input variable unit of measure

Commit:
fix unit test

Commit:
remove unused constants from the previous HI method

Commit:
Merge remote-tracking branch 'origin/develop' into extendedHI

Commit:
remove extended HI output (it replaces the old one)

yujiex commented:
Hi @Myoldmopar and @mjwitte this is ready for review. Thanks.

Myoldmopar commented on doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex:
```diff
@@ -66,69 +66,17 @@ \subsubsection{Heat Index}\label{heat-index}
-regression is not valid for extreme temperature and relative humidity conditions
-beyond the range of data considered by Steadman.
+Starting from version 24.2, the heat index calculation adopts the extended heat
+index method developed by Lu & Romps [17]. The previous heat index gives
```
There is a LaTeX error here with the & character, it needs to be escaped.

yujiex replied:
Thanks. I just escaped it.

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -262,6 +262,8 @@ set(SRC
     ExhaustAirSystemManager.cc
     ExhaustAirSystemManager.hh
+    extendedHI.cc
+    extendedHI.hh
```
I don't like the lower cased names...  I mean in general I am fine with it, but it looks super weird against all the other PascalCased names.  I'm going to change this if I push changes up.

yujiex replied:
Thanks. I just changed the filenames to PascalCased names.

Myoldmopar commented on src/EnergyPlus/extendedHI.cc:
```diff
@@ -0,0 +1,369 @@
+
+        if (show_info) {
+            if (region == "I") {
+                std::cout << "Region I, covering (variable phi)\n";
```
Are these supposed to be debug messages that can be removed?  I dislike seeing std::cout directly in E+ code, even if it's only meant to be used during debugging.

yujiex replied:
They are debugging messages. I will remove them.

Myoldmopar commented on src/EnergyPlus/extendedHI.cc:
```diff
@@ -0,0 +1,369 @@
+        // Dictionary to map eqvar_name to tuple index
+        std::map<std::string, int> dic = {{"phi", 1}, {"Rf", 2}, {"Rs", 3}, {"Rs*", 3}, {"dTcdt", 4}};
+        auto eqvars = find_eqvar(state, Ta, RH);
+        std::string eqvar_name = std::get<0>(eqvars);
```
Are you asking a function to return a string, and then doing a string comparison to decide which code path to follow?  For one thing, this arrangement seems way way overkill just to calculate heat index.  And second, it cannot be a string, you need to _at least_ use an enum.  But I really feel this whole thing could be way simpler.

yujiex replied:
I will simplify this

Myoldmopar commented on src/EnergyPlus/extendedHI.hh:
```diff
@@ -0,0 +1,77 @@
+} // namespace extendedHI
+} // namespace EnergyPlus
+
+#endif
```
No newline here.

yujiex replied:
Thanks, I just added newline at the end.

Myoldmopar commented on tst/EnergyPlus/unit/extendedHI.unit.cc:
```diff
@@ -0,0 +1,734 @@
+                                  27187.715714871047,
+                                  41643.766112228834,
+                                  62053.2640569111,
+                                  90163.72448626769};
```
I'm inclined to say all the values in this unit test are _waaay_ too precise.  We cannot know these values so precisely, so to test against them is futile.  I see you have tolerance set to e-8, which is pretty tight here.  But there are other ones where tolerance is more like e-4, yet the values you compare against have like 12+ decimals.  It's not the end of the world, it's just a lot of numbers here :)

yujiex replied:
I will unify the tolerance and the values here.

These are output from this python file. I was trying to ideally match their result. 
[heatindex.py.zip](https://github.com/user-attachments/files/16146813/heatindex.py.zip)

The e-4 part is due to the difference in the root solvers between EnergyPlus and the source python file  (when EnergyPlus uses the bisection one as well).
The following is a illustration of the difference in a somewhat extreme case
![image](https://github.com/NREL/EnergyPlus/assets/13421161/6697fd07-a6b1-455d-8e89-321536043d2f)
Here is the difference in the code
<img width="1216" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/d4b9a017-49cf-453a-be3c-449580395849">

Myoldmopar commented on tst/EnergyPlus/unit/extendedHI.unit.cc:
```diff
@@ -0,0 +1,734 @@
+TEST_F(EnergyPlusFixture, extendedHI_heatindex)
+{
+
+    // fixme: not finished
```
Fix me

Myoldmopar commented on src/EnergyPlus/extendedHI.cc:
```diff
@@ -0,0 +1,369 @@
+        // show_info: whether to print some messages. Might be useful in debugging
+        // The function computes the extended heat index, in Kelvinn
+
+        auto const HVACSystemRootSolverBackup = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
```
Ummmm making a copy of the whole root solver object each time we calculate heat index.... that's a no.

yujiex replied:
I see. I changed it to this

```
auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverAlgorithm::Bisection;
// do stuff
state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverBackup;
```

Myoldmopar reviewed (changes requested):
OK, so my main concern is around the complexity of the heat index calculation.  Asking functions to return pairs and tuples that include strings, and then checking the string to do different operations, is all .... a lot.  I'm also highly concerned about making a copy of the entire root solver object.  This needs some work.  Simplification work.

I have less critical other concerns: a LaTeX error, the filename should ideally be ExtendedHI.* so that it matches our existing pattern, the unit test precision gives some feels but might be fine.

Commit:
modify latex escape char

Commit:
rename files with PascalCased names

Commit:
rename unit test file to PascalCased

Commit:
removing debugging print code

Commit:
change namespace to PascalCased, add newline EOF

Commit:
remove string output of find_T

Commit:
simplify find_eqvar, find_T, adjust precision of unit tests

Commit:
fix enum format, clang-format, include filename

Commit:
more clang-format

Commit:
prevent creating a rootsolver object while saving old value

yujiex commented:
Hi @Myoldmopar, I've simplified the functions some more and addressed other comments as well. Please see if they look okay. Thanks.

Commit:
Merge remote-tracking branch 'origin/develop' into extendedHI

Commit:
fix CMakeList unit test file name

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,499 @@
+        // The function computes the extended heat index, in Kelvinn
+
+        auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
+        auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
```
This is still making a copy of the root solver each call to heatIndex.  Which happens for each zone at each time step.  This is not OK.  I get the feeling you shouldn't use the HVACSystemRootSolver for this.  Instead just use SolveRoot?  

As an aside, I still think the find_eqvar with tuple and returning different things is still a bit too complex just to calculate heat index.  At this point I wouldn't hold it up just for that, but if it could be simplified further, that would be great.

yujiex replied:
Hi Edwin, I was using `General::SolveRoot` to solve the roots. The purpose of the part involving `HVACSystemRootSolver` is just so that the root searching order inside the `General::SolveRoot` function follows the bisection method, which is what was used in the original python implementation. Maybe I'll just remove this part and hopefully they don't differ much when using different root finding schemes. So far the tests in the unit test file still passes if this part is removed.

I will try to simplify `find_eqvar` some more

yujiex replied:
@Myoldmopar I just changed `find_eqvar` to two functions each returning a single value.

yujiex replied:
btw I also just printed out the size of the things involved here. They're all 4 byte.
```
auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
```
![image](https://github.com/user-attachments/assets/bb863447-27b9-4749-91b4-c7605112fec8)

yujiex replied:
Removing the 
```
        auto const &HVACSystemRootSolver = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
        auto const HVACSystemRootSolverBackup = HVACSystemRootSolver;
        state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverAlgorithm::Bisection;

...
        state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverBackup;
```
seems to cause some NaN in the unit test `HeatBalanceSurfaceManager_TestResilienceMetricReport`, when calculating this: `ExtendedHI::heatindex(state, 298.15, 0.5) `. Also in unit test `HeatBalanceSurfaceManager_TestThermalResilienceReportRepPeriod`

Myoldmopar reviewed (changes requested):
This is still making a copy of the root solver object in each call to heat index, and that needs to be addressed before merging.  If you are needing to simply find a zero, you can do this which is what we do in a ton of places:

```
int constexpr maxIter = 1000;
Real64 constexpr xMin = -10.0;
Real64 constexpr xMax = 10.0;
Real64 constexpr accuracy = 0.001;
Real64 someTargetCondition = abc;
auto f = [&state] (Real64 x) { return x - someTargetCondition; };
int flag = 0;
Real64 x = 0.0;
General::SolveRoot(state, accuracy, maxIter, flag, x, f, xMin, xMax);
```

Is it possible to use that here rather than the HVACSolver stuff?

Commit:
remove setting root solver root finding scheme to bisection

the removed code was trying to set the root finding scheme,
HVACRootSolverALgorithm::Bisection, following the original python
implementation. This is removed to prevent copying the object

Commit:
simplify find_eqvar to return name and value separately

Myoldmopar commented:
@yujiex I apologize for sending you down a distraction path here.  When I saw that you were making a copy of 
`state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver` I immediately thought it was making a copy of a root solver.  Like a whole root solver.  Since that's the name of the variable :) 

When you said it was only 4 bytes, that of course made things different.  It's just a freaking enum.  Sorry for distracting you with this, please revert any unnecessary changes and let's get this merged.

Commit:
Revert "remove setting root solver root finding scheme to bisection"

This reverts commit 17341b0393ae55a7db3863841b587e8d05f11cc2.

Commit:
Revert "prevent creating a rootsolver object while saving old value"

This reverts commit 7f824c13f8b929da65850782af1ff1f2681acc7c.

yujiex commented:
> @yujiex I apologize for sending you down a distraction path here. When I saw that you were making a copy of `state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver` I immediately thought it was making a copy of a root solver. Like a whole root solver. Since that's the name of the variable :)
> 
> When you said it was only 4 bytes, that of course made things different. It's just a freaking enum. Sorry for distracting you with this, please revert any unnecessary changes and let's get this merged.

It's totally fine. It might be more clear if the variable was named something like RootSearchingMode. Currently it sounds like very memory intensive.

I just reverted the changes.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into extendedHI

Myoldmopar commented:
Pulling develop in here as well to get fresh CI results.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into extendedHI

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,538 @@
+    constexpr Real64 epsilon = 0.97;                  //           , emissivity of surface, steadman1979
+    constexpr Real64 M = 83.6;                        // kg        , mass of average US adults, fryar2018
+    constexpr Real64 H = 1.69;                        // m         , height of average US adults, fryar2018
+    Real64 A = 0.202 * pow(M, 0.425) * pow(H, 0.725); // m^2       , DuBois formula, parson2014
```
This should be constexpr-able right?  And I will assume for now that these are indeed used in multiple locations where it makes sense to keep them out here in the namespace.

yujiex replied:
yes. It should be. Yeah, some stuff like `Pc` is used in a few functions.

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,538 @@
+    {
+        constexpr Real64 hc = 17.4;
+        constexpr Real64 phi_rad = 0.85;
+        Real64 hr = epsilon * phi_rad * sigma * (std::pow(Ts, 2) + std::pow(Ta, 2)) * (Ts + Ta);
```
In places where you can't `constexpr`, it's still good to `const`.

yujiex replied:
I see. Good point

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,538 @@
+        Real64 m = (Pc - Pa) / (Zs(Rs) + Za);
+        Real64 Ts;
+        int SolFla;
+        General::SolveRoot(
```
Yes, this is the way.  If you want to make it slightly clearer, you _can_ define the lambda first so that you have plenty of room to name things nicely, and then just pass the lambda in as the argument.  It's fine as an argument here though.

yujiex replied:
Make sense. Probably better as their own function in front

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,538 @@
+                Tf,
+                [&](Real64 Tf) {
+                    return (Tf - Ta) / Ra_bar(Tf, Ta) + (Pc - Pa) * (Tf - Ta) / ((Zs(Rs) + Za_bar) * (Tf - Ta) + r * Ra_bar(Tf, Ta) * (Ts_bar - Tf)) -
+                           (Tc - Ts_bar) / Rs;
```
This is definitely getting pretty long to just be a lambda argument.  I'd consider making this a nicely named function with clear arguments, etc., and then just have the lambda call that function.  Debugging an issue with the arithmetic in this line would be quite difficult.

yujiex replied:
Yeah, indeed. This is probably better separated out.

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,538 @@
+        if (Ta == 0.0) T = 0.0;
+
+        state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver = HVACSystemRootSolverBackup;
+        return T;
```
In general, this is so much better than it was, great job there.  I am still quite surprised to find the sheer amount of SolveRoot calls inside this module which is just trying to calculate heat index.  Anyway, thanks for the cleanups.

yujiex replied:
Thanks for reviewing. Glad I've been able to clean it out a bit. It does seem a pretty involved calculation process in their paper as well.

Myoldmopar commented on src/EnergyPlus/ExtendedHI.hh:
```diff
@@ -0,0 +1,91 @@
+        Rf = 2,
+        Rs = 3,
+        DTcdt = 4,
+        Num
```
I'm not convinced Invalid and Num are needed here.  This enum is not paired with any input strings, right?  I would suggest removing Invalid and Num, moving this enum class into the cc file (since it's not used here in the header), and then adding this enum to the exclude list [here](https://github.com/NREL/EnergyPlus/blob/d6e49668d138eea326cc7917b75215fbca7ecf64/scripts/dev/check_for_malformed_enums.py#L109) and [here](https://github.com/NREL/EnergyPlus/blob/d6e49668d138eea326cc7917b75215fbca7ecf64/scripts/dev/check_for_malformed_enums.py#L132)

Myoldmopar commented on tst/EnergyPlus/unit/ExtendedHI.unit.cc:
```diff
@@ -0,0 +1,625 @@
+        }
+    }
+    HI = (HI - 32.0) * (5.0 / 9.0);
+    return HI;
```
Interesting.  I guess this is fine.  I start getting nervous when the amount of logic inside the unit tests grows.  Like now I need to test this function as well...  But I get the intent here, so I'll just move on.

Myoldmopar commented on tst/EnergyPlus/unit/ExtendedHI.unit.cc:
```diff
@@ -0,0 +1,625 @@
+constexpr Real64 Fahrenheit2Celsius(Real64 F)
+{
+    return (F - 32.0) * 5.0 / 9.0;
+}
```
I feel like we definitely have this somewhere else so you don't need your own.

Myoldmopar commented on tst/EnergyPlus/unit/ExtendedHI.unit.cc:
```diff
@@ -0,0 +1,625 @@
+                                                   377.4622562734294,
+                                                   381.56900731119094,
+                                                   385.6185252717114,
+                                                   389.61199074954493}};
```
So where did these values come from?  I am guessing they are just generated from running the code, not from an external source.  If so, this is OK, but not great.  If these values are from a separately validated source, then that is a nice verification.

Myoldmopar reviewed (approved):
I'm not requesting any specific changes.  I left some comments that you can respond to, but I think this is probably fine to go in.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into extendedHI

yujiex commented:
> I'm not requesting any specific changes. I left some comments that you can respond to, but I think this is probably fine to go in.

Thanks, @Myoldmopar. I will address these today

Myoldmopar commented:
I pushed the branch with develop pulled in and will wait for CI to give fresh results.  Once that's done, this can merge.  @yujiex if you want to address any of my comments via code changes, they can come later.  For now just respond to them as comments however much you want.

yujiex commented:
> I pushed the branch with develop pulled in and will wait for CI to give fresh results. Once that's done, this can merge. @yujiex if you want to address any of my comments via code changes, they can come later. For now just respond to them as comments however much you want.

I see. Thanks @Myoldmopar

Myoldmopar commented:
Still happy here locally with develop pulled in, merging.  Thanks @yujiex

yujiex commented:
Thanks @Myoldmopar !
